### PR TITLE
RichText: Fix delete key in empty contenteditables

### LINKED ISSUE #5345: Regression: Delete key doesn't work in empty paragraph block
Steps to reproduce:

- Write text
- Create linebreak between two paragraphs
- Press Delete

Note that nothing happens. In 2.2 the empty paragraph block would be deleted. 

delete-doesn t-work

### PR DESCRIPTION
**Pull request by youknowriad**
closes #5345 related #5159

On empty content editables, there's always a `<br />` element added by TinyMCE, this makes `isHorizontalEdge` return false where it should return `true`. This fixes this issue.

Also, another issue I found related to #5159 is that we want to trigger onSplit only if we click delete at the end of RichText or backspace at the beginning. So we need to check both the vertical and the horizontal edge in both cases.

**Testing instructions**

 - Try backspacing or clicking delete on empty paragraph, at the beginning, end, middle of paragraphs. It should behave properly.

### CODE REVIEW

**Comment by jasmussen** [comment]
You ROOOOOCK! 🗿 

Heck yes this works. I can't break this! 🚢

File: blocks/rich-text/index.js [hunk 1]
```diff
@@ -478,8 +479,8 @@ export class RichText extends Component {
 		const rootNode = this.editor.getBody();
 
 		if (
-			( event.keyCode === BACKSPACE && isHorizontalEdge( rootNode, true ) ) ||
-			( event.keyCode === DELETE && isHorizontalEdge( rootNode, false ) )
+			( event.keyCode === BACKSPACE && isVerticalEdge( rootNode, true ) && isHorizontalEdge( rootNode, true ) ) ||
```
Comment (iseulde) [inline blocks/rich-text/index.js]: > Also, another issue I found related to #5159 is that we want to trigger onSplit only if we click delete at the end of RichText or backspace at the beginning. So we need to check both the vertical and the horizontal edge in both cases.

Huh, horizontal edge is only supposed to check  the first and last caret position? Not every horizontal line edge? So I don't understand why this is needed?

Comment (youknowriad) [reply blocks/rich-text/index.js]: Oh I see, this was not clear for me. updated

File: utils/dom.js [hunk 2]
```diff
@@ -58,7 +58,12 @@ export function isHorizontalEdge( container, isReverse, collapseRanges = false )
 
 	const maxOffset = node.nodeType === TEXT_NODE ? node.nodeValue.length : node.childNodes.length;
 
-	if ( ! isReverse && offset !== maxOffset ) {
+	if (
+		! isReverse &&
+		offset !== maxOffset &&
+		// content editables with only a BR element are considered empty
+		( offset !== 0 || maxOffset !== 1 || node.nodeType === TEXT_NODE || ! node.childNodes.length || node.childNodes[ 0 ].tagName !== 'BR' )
```
Comment (iseulde) [inline utils/dom.js]: I find this whole thing quite hard to follow. Would it be possible the separate this case? Also, this seems to look for `BR` deeply nested too, while it seems we might just want the direct child of the container?

Comment (iseulde) [reply utils/dom.js]: I wonder if we can add an `isEmpty` helper, check for that with the `container` and return true early.

Comment (iseulde) [reply utils/dom.js]: Hm, I wonder why this worded previously... https://github.com/WordPress/gutenberg/pull/5159/files
No br checks .

Comment (iseulde) [reply utils/dom.js]: Most likely because MCE corrects wrong ranges.

Comment (youknowriad) [reply utils/dom.js]: Updated to rely on TinyMCE API instead.

**Comment by youknowriad** [comment]
Seems like there's a GitHub bug because the travis build is successful

**Comment by aduth** [comment]
The behavior on master produces an error. Resolving the error seems to resolve the issue without any other changes.

Applied on master:

```diff
diff --git a/utils/dom.js b/utils/dom.js
index c90b4c7a0..bf2f96bd6 100644
--- a/utils/dom.js
+++ b/utils/dom.js
@@ -65,7 +65,7 @@ export function isHorizontalEdge( container, isReverse, collapseRanges = false )
 	while ( node !== container ) {
 		const parentNode = node.parentNode;
 
-		if ( parentNode[ `${ order }Child` ] !== node ) {
+		if ( parentNode && parentNode[ `${ order }Child` ] !== node ) {
 			return false;
 		}
``

**Comment by aduth** [comment]
Above only seems to fix standard backspace, not forward delete.

**Comment by aduth** [comment]
It'd be nice to have some E2E tests for these behaviors.

**Comment by youknowriad** [comment]
Standard backspace work for me on master, it's probably another bug. How to reproduce your issue?

**Comment by aduth** [comment]
1. Create two paragraph blocks
2. Press enter after the first paragraph block
3. Press [Windows: Delete, Mac: Ctrl+Delete]

Expected: Blank paragraph merges into the last paragraph, with caret at start of last paragraph
Actual: Blank paragraph merges into first paragraph, with caret at end of first paragraph†

† In earlier testing, I swear this was doing _nothing at all_.

File: utils/dom.js [hunk 3]
```diff
@@ -57,8 +58,13 @@ export function isHorizontalEdge( container, isReverse, collapseRanges = false )
 	}
 
 	const maxOffset = node.nodeType === TEXT_NODE ? node.nodeValue.length : node.childNodes.length;
-
-	if ( ! isReverse && offset !== maxOffset ) {
+	const editor = node.nodeType !== TEXT_NODE && tinymce.get( node.getAttribute( 'id' ) );
```
Comment (aduth) [inline utils/dom.js]: Minor: `id` is a property of `Element`:

```js
node.id
```

https://developer.mozilla.org/en-US/docs/Web/API/Element/id

**Comment by aduth** [comment]
Should we just revert #5159? At least the parts touching `RichText`?

**Comment by youknowriad** [comment]
@aduth It's an option, no strong opinion for me.

**Comment by youknowriad** [comment]
Do you think reverting (specific behavior in RichText) is better than having this change in the way `isHorizontalEdge` work?

**Comment by aduth** [comment]
I think we're safer in assuming that it works correctly, as #5159 was meant largely as a refactoring simplification. I'm not too confident that the new logic works correctly, as we apparently don't have a good understanding of transformations applied to `editor.selection.getRng` vs. `window.getSelection().getRangeAt`.

I'm also motivated to be consistent in how we're using `tinymce` in `utils/dom`. Either we shouldn't use it at all, or we should use it over things like `window.getSelection` etc.

**Comment by youknowriad** [comment]
Fair enough restored. Another option we could explore as a follow up is to rely on the `tinymce` selection object if we detect that the `node` is a tinymce editor.

**Comment by youknowriad** [comment]
It looks like relying on `isStartOfEditor` and `isEndOfEditor` was not working properly sometimes:

1. Write two paragraphs.
2. Move the cursor to the middle of the *first* paragraph, then hit Right until you’re at the edge.
3. Press Delete: the merge happens properly.
4. Repeat 1. Place the caret in the middle of the *second* paragraph, then press Left until you’re at the edge. Press Left once more to move to the edge of the first paragraph.
5. Press Delete: the merge fails

I restored my original fix to keep relying on `isHorizontalEdge` it should be slightly better even if the caret is not at the right position after the instructions above. But I think it's ok for now, it's same on master.

**Comment by mcsf** [comment]
The last state of this PR seems to fix the issue pretty well!

**Review by mcsf — APPROVED** [review]
LGTM!
